include/nuttx/spinlock.h

```diff
@@ -529,6 +529,7 @@ irqstate_t spin_lock_irqsave_wo_note(FAR volatile spinlock_t *lock)
 {
   irqstate_t ret;
   ret = up_irq_save();
+  sched_lock_wo_note();
```

Comment:
If a non-preemptive version of spinlock_irqsave is implemented, We could disable global interrupts after spinlock is acquired, which will allow the system to respond interrupts during the test set loop and improve RT performance

Reply:
@anchao 
I'm not entirely sure if I understand it correctly,
but if we disable interrupts after acquiring a spinlock,
when an interrupt occurs during the execution of a test-and-set loop in spin_lock_irqsave
and attempts to acquire the same spinlock within the interrupt handler, it can lead to a deadlock.
Synchronization between interrupts and threads cannot be achieved in this case.

Reply:
Yes, you are right. it will trigger deadlock If access the spinlock in interrupt handler, but RT model should reduce the interrupt disable phase as much as possible. I think it may be necessary to implement a set of independent APIs to prohibit developers from using them in interrupt handler, this might be another topic.

Reply:
spinlock_irqsave is used only when you need sync with interrupt handle, spinlock_lock is enough for normal thread protection. So, the sequence must be disabling interrupt and try hold the lock, to avoid disabling the interrupt too long, the compromised approach is:
```
for (; ; )
  {
    ret = up_irq_save();
    if (up_testset() == 0)
      {
         break;
      }
    up_irq_restore(ret);
  }
```